src/Analyser/NodeScopeResolver.php

```diff
@@ -2633,7 +2635,8 @@ static function (Node $node, Scope $scope) use ($nodeCallback): void {
 					$filteringExpr = new BinaryOp\BooleanOr($filteringExpr, $armCondExpr);
 				}
 
-				$armNodes[] = new MatchExpressionArm($condNodes, $arm->getLine());
+				$matchArmBody = new MatchExpressionArmBody($armCondScope, $arm->body);
```

Comment:
I think the correct scope here is the one that's created below in processExprNode call - `$matchScope->filterByTruthyValue($filteringExpr)` but please test that :)